### PR TITLE
Mention existence of various undocumented tactics

### PR DESCRIPTION
**Pull request by jfehrle**
We can mark some of these as "for internal use only", but for completeness we should mention that they exist.

Attn: @Zimmi48

### CODE REVIEW

**Comment by Zimmi48** [comment]
@coq/micromega-maintainers @coq/nsatz-maintainers @coq/ring-maintainers @coq/funind-maintainers @coq/tactics-maintainers Could you please have a look at the undocumented tactics that are being added to the reference manual in this PR and tell us if they should be marked "for internal use only" or if they should be properly documented (in which case, could you provide us some short description for those)?

**Comment by Matafou** [comment]
I think `soft functional induction` was mainly used for debugging. Ping @forestjulien for verification.

**Comment by ppedrot** [comment]
The ring-related tactics are really internal. (And they should be ultimately removed in favour of something slightly more sensible.)

**Comment by jfehrle** [comment]
It's better to mark them as "for internal use" because a) users may encounter them and b) because there is tooling to identify undocumented tactics, so the question of whether a tactic in for internal use will come up repeatedly without a mechanism to identify them as such in the code.  It's not worth the trouble  to create such a mechanism.

**Comment by ppedrot** [comment]
> It's not worth the trouble to create such a mechanism.

I can't disagree more. The more we introduce programming in tactic languages, the more we're going to face issues with internal functions that should never have been exposed to the user in the first place. It does *not* make sense at all to document these functions because it's enshrining the fact we suck at abstraction. Otherwise why don't you just pick every single entry in the cmx files and declare that this is part of the API while we're at it?

The basic mechanism in ML is to have module interfaces, but due to the way modules work in Coq (i.e. LIBOBJECT :ghost:) it's very hard to properly encapsulate non-logical objects. I didn't manage to get anything in Ltac2 for instance in the current architecture.

(There was an issue about private tactics for Ltac1 but I don't find it right now.)

**Comment by Zimmi48** [comment]
> It's better to mark them as "for internal use" because a) users may encounter them and b) because there is tooling to identify undocumented tactics, so the question of whether a tactic in for internal use will come up repeatedly without a mechanism to identify them as such in the code. It's not worth the trouble to create such a mechanism.

I didn't really take the time to write this because IMO this was not up for debate: the rule is simple, document everything that is user accessible even if it is to tell that it is internal and should not be used. Everything that is publicly accessible is part of the API, whether developers want it or not. That is a basic fact from software engineering. It's better to document things as "should not be relied on" rather than not documenting them at all.

Then, obviously we should design better module / encapsulation mechanisms. But making them documentation-specific would make no sense.

**Comment by SkySkimmer** [comment]
Not for 8.15

**Comment by jfehrle** [comment]
I marked the tactics added in ring and micromega as "For internal use only".

At this point, do we prod others for answers or simply proceed?

File: doc/sphinx/proofs/automatic-tactics/auto.rst
Comment (Zimmi48) [inline doc/sphinx/proofs/automatic-tactics/auto.rst:142]: You can document this one as an alias for `eauto`.

**Comment by Zimmi48** [comment]
@forestjulien Can you confirm that  `soft functional induction` should be marked as internal (not for use)?

@coq/nsatz-maintainers What about `nsatz_compute`?

@mattam82 What about `autounfold_one`?

Anyone: what about `gintuition` / `simple injection`?

**Comment by jfehrle** [comment]
Updated.

**Comment by herbelin** [comment]
> Anyone: what about `gintuition` / `simple injection`?

Afaiu, `gintuition` is to `intuition` what `firstorder` is to `tauto`. It is worth to be documented I would say.

`simple injection` is a priori useful in Ltac's code as it put equations quantified on the goal. However, I found only one occurrence in my collection of Coq developments, so maybe it is not necessary to keep it.

**Comment by Zimmi48** [comment]
> Afaiu, `gintuition` is to `intuition` what `firstorder` is to `tauto`. It is worth to be documented I would say.

Given that `firstorder` also accepts an `ltac_expr` argument and uses `auto with core` by default, does this mean that `gintuition` with no argument is equivalent to `firstorder idtac` and that `gintuition` with an argument is the same as `firstorder` with the same argument?

**Comment by Zimmi48** [comment]
> `simple injection` is a priori useful in Ltac's code as it put equations quantified on the goal. However, I found only one occurrence in my collection of Coq developments, so maybe it is not necessary to keep it.

Actually, `injection` without the `as` clause still does this unless the `Structural Injection` flag is on. So I guess `simple injection` is useful mostly to the rare users that have turned this flag on. Worth keeping, I'd say.

File: doc/sphinx/proofs/automatic-tactics/logic.rst
Comment (jfehrle) [reply doc/sphinx/proofs/automatic-tactics/logic.rst:91]: Got the first part, but who is the second paragraph addressed to?

**Comment by jfehrle** [comment]
Updated.

File: doc/sphinx/proofs/automatic-tactics/logic.rst
Comment (Zimmi48) [reply doc/sphinx/proofs/automatic-tactics/logic.rst]: To @herbelin

File: doc/sphinx/proofs/automatic-tactics/auto.rst
Comment (Zimmi48) [inline doc/sphinx/proofs/automatic-tactics/auto.rst:139]: BTW, now that we've removed `bfs eauto`, we should deprecate `dfs eauto`. It is probably unused, and it doesn't particularly make sense to keep it around.

File: doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst
Comment (Zimmi48) [inline doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:654]: From Hugo's explanation, you can add a description for this tactic (and make it a variant of `injection`) saying something like: `simple injection` will always add the derived equalities as new premises in the goal instead of putting them in the context, even if the `Structural Injection` flag is on.

**Comment by Zimmi48** [comment]
> @forestjulien Can you confirm that `soft functional induction` should be marked as internal (not for use)?
> 
> @coq/nsatz-maintainers What about `nsatz_compute`?
> 
> @mattam82 What about `autounfold_one`?

Last call. In all cases, we will list these tactics in the manual, even if we leave them undocumented for now.

File: doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst
Comment (jfehrle) [reply doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:654]: Here's what I have now.  I also reworded the description of the flag, dropping "erases the original hypothesis" because I expect `injection` always does that; the flag doesn't change that.  Is that correct?

   .. tacn:: simple injection {? @induction_arg }

      Similar to :tacn:`injection`, but always adds the derived equalities
      as new :term:`premises <premise>` in the current goal (instead of as
      new hypotheses) even if the :flag:`Structural Injection` flag is set. 

   .. flag:: Structural Injection

      When set, :n:`injection @term` adds the generated equalities as
      new hypotheses rather than adding them to the current goal as
      :term:`premises <premise>`, as if giving :n:`injection @term as`
      (with an empty list of names). This flag is off by default.

**Comment by jfehrle** [comment]
Updated.

File: doc/sphinx/proofs/automatic-tactics/auto.rst
Comment (jfehrle) [reply doc/sphinx/proofs/automatic-tactics/auto.rst:139]: See #15379

File: doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst
Comment (Zimmi48) [inline doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:638]: It would be better to keep the glossary definition: "When this `flag` is set, ..."

File: doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst
Comment (Zimmi48) [reply doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst]: Actually no, it only erases the original hypothesis if `as` is used or `Structural Injection` is set...

File: doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst
Comment (jfehrle) [reply doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst:638]: i.e. to be clear it doesn't refer to "injection ...",  Sure.

File: doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst
Comment (jfehrle) [reply doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst]: OK, then

When this :term:`flag` is set, :n:`injection @term` erases the original hypothesis
      and adds the generated equalities as new hypotheses rather than adding them
      to the current goal as :term:`premises <premise>`, as if giving :n:`injection @term as`
      (with an empty list of names). This flag is off by default.

**Comment by jfehrle** [comment]
Updated.

**Review by Zimmi48 — APPROVED** [review]
I will merge this early next week if there are no more comments (in particular, no more suggestions on how to document `nsatz_compute`, `autounfold_one`, `soft functional induction`).

**Comment by Zimmi48** [comment]
@jfehrle A rebase is needed now that #15379 is merged.

**Comment by jfehrle** [comment]
Rebased and squashed.

**Review by Alizter — APPROVED** [review]

**Comment by Zimmi48** [comment]
@coqbot merge now

**Pull request merged**
